crates/sqllib/src/casts.rs

```diff
@@ -14,31 +19,113 @@ use num::{FromPrimitive, One, ToPrimitive, Zero};
 use num_traits::cast::NumCast;
 use regex::{Captures, Regex};
 use rust_decimal::{Decimal, RoundingStrategy};
-use std::collections::BTreeMap;
-use std::error::Error;
-use std::string::String;
+use std::{cmp::Ordering, collections::BTreeMap, error::Error, string::String};
+
+#[doc(hidden)]
+pub(crate) fn type_name(name: &'static str) -> &'static str {
+    match name {
+        "i8" => "TINYINT",
+        "i16" => "SHORTINT",
+        "i32" => "INTEGER",
+        "i64" => "BIGINT",
+        "f" => "REAL",
+        "d" => "FLOAT",
+        "Timestamp" => "Timestamp",
+        "Date" => "DATE",
+        "Time" => "TIME",
+        "decimal" => "DECIMAL",
+        "ShortInterval" => "INTERVAL",
+        "LongInterval" => "INTERVAL",
+        "s" => "(VAR)CHAR",
+        _ => "Unexpected type",
+    }
+}
+
+#[doc(hidden)]
+pub(crate) fn ttype_name(name: &'static str) -> &'static str {
+    match name {
+        "i8" => "TINYINT",
+        "i16" => "SHORTINT",
+        "i32" => "INTEGER",
+        "i64" => "BIGINT",
+        "F32" => "REAL",
+        "F64" => "FLOAT",
+        "Timestamp" => "Timestamp",
+        "Date" => "DATE",
+        "Time" => "TIME",
+        "Decimal" => "DECIMAL",
+        "ShortInterval" => "INTERVAL",
+        "LongInterval" => "INTERVAL",
+        "String" => "(VAR)CHAR",
+        _ => "Unexpected type",
+    }
+}
+
+#[macro_export]
+macro_rules! tn {
+    ($result_name: ident) => {
+        type_name(stringify!($result_name))
+    };
+}
+
+#[macro_export]
+macro_rules! ttn {
+    ($result_type: ty) => {
+        ttype_name(stringify!($result_type))
+    };
+}
+
+macro_rules! cn {
+    ($result_name: ident) => {
+        cast_null(tn!($result_name))
+    };
+}
+
+#[doc(hidden)]
+pub fn unwrap_cast<T>(value: SqlResult<T>) -> T {
+    match value {
+        Err(ce) => panic!("{}", *ce),
+        Ok(v) => v,
+    }
+}
+
+#[doc(hidden)]
+pub fn unwrap_safe_cast<T>(value: SqlResult<Option<T>>) -> Option<T> {
+    value.unwrap_or_default()
+}
+
+fn cast_null(t: &str) -> Box<SqlRuntimeError> {
+    SqlRuntimeError::from_string(format!("cast of NULL value to non-null type {}", t))
+}
 
 // Creates three cast functions based on an existing one
-macro_rules! cast_function {
+// The original function is
+// cast_to_ $result_name _ $type_name( value: $arg_type ) -> SqlResult<$result_type>
+macro_rules! cast_func {
     ($result_name: ident, $result_type: ty, $type_name: ident, $arg_type: ty) => {
         ::paste::paste! {
             #[doc(hidden)]
             #[inline]
-            pub fn [<cast_to_ $result_name N_ $type_name>]( value: $arg_type ) -> Option<$result_type> {
-                Some([<cast_to_ $result_name _ $type_name>](value))
+            pub fn [<cast_to_ $result_name N_ $type_name>]( value: $arg_type ) -> SqlResult<Option<$result_type>> {
+                r2o([<cast_to_ $result_name _ $type_name>](value))
```

Comment:
r2o means relation to object?

Reply:
it's Result to Option. I thought it's obvious from the definition.